justinmahar/github-projectv2-csv-exporter
# feat: support custom project fields

# Summary
- Resolves #5 
- Introduces two new features:
	- Display custom GH project fields in generated CSV output
	- Toggle all CSV headers on/off

# Details
I had to make quite a few changes in order to get this feature to work- feel free to request modifications if anything seems questionable or otherwise unacceptable:
- A new component (`GitHubProjectFieldSettings.tsx`) is added to contain the project field settings logic
- A new API method (`fetchProjectFields()`) is added to query the names (and types) of all project item fields
- I swapped out the old csv generation package [export-to-csv](https://github.com/alexcaza/export-to-csv) for [json-2-csv](https://github.com/mrodrig/json-2-csv); not only is it more up-to-date, but it also supports setting custom headers, which the old library did not have sufficient customization for
- Slightly more React state is used throughout the app, but I felt no noticeable overhead in my local env

# Testing
- Clone fork, checkout `crimlog` branch
- `npm install`
- `npm start`
- Navigate to http://localhost:6006 in your browser and proceed to [Exporter Settings](http://localhost:6006/?path=/story/tools-github-project-exporter--settings)
	- Set an access token and login value of "crimlog" (or your own project)
- On the [Exporter Settings](http://localhost:6006/?path=/story/tools-github-project-exporter--settings) page, you should be able to view several different fields as the last setting, along with the ability to toggle them on/off 
- Run an actual export in a local environment and confirm the actual CSV output matches the expected result

# Imagery
Web application:
![image](https://user-images.githubusercontent.com/35435704/207775366-1abb5c6e-063f-40ef-8b4f-ff09f9475f50.png)

CSV output:
![image](https://user-images.githubusercontent.com/35435704/207772883-1877c3a6-62a1-4e4e-a836-3db2e52de852.png)
